### PR TITLE
server: reformat/refactor around redirects (again)

### PR DESCRIPTION
**Pull request by jorgeorpinel**
Second attempt at #973.

### CODE REVIEW

**Comment by jorgeorpinel** [comment]
Weird. This works for me locally (`yarn dev`) but crashes in the (built) review app (https://dvc-landing-server-redi-btalyj.herokuapp.com/)... Investigating ⏳

**Comment by jorgeorpinel** [comment]
@iAdramelk @shcheklein related: do you guys know why when you `yarn build` and `yarn start` this locally, it tries to enforce HTTPS on localhost:3000? Makes it impossible for me to debug the built app locally.

**Comment by jorgeorpinel** [comment]
Anyway, looking at https://dashboard.heroku.com/apps/dvc-landing-server-redi-btalyj/logs I see the review app error log is showing this a lot:

```
2020-02-20T07:43:44.228837+00:00 app[web.1]: TypeError: Cannot read property 'amp' of null
2020-02-20T07:43:44.228839+00:00 app[web.1]:     at Server.findPageComponents (/app/node_modules/next/dist/next-server/server/next-server.js:465:19)
...
```

which unfortunately is not a very informative stack trace... But per https://github.com/zeit/next.js/issues/7303#issuecomment-511053550 this probably has to do with my changes around `url.parse`... Will try one of the fixes in that report ⏳

**Comment by iAdramelk** [comment]
@jorgeorpinel 

> related: do you guys know why when you yarn build and yarn start this locally, it tries to enforce HTTPS on localhost:3000? Makes it impossible for me to debug the built app locally.

You can temporarily disable this redirect by commenting these lines: https://github.com/iterative/dvc.org/blob/master/src/utils/redirects.js#L37-L39

**Comment by shcheklein** [comment]
@iAdramelk @fabiosantoscode @jorgeorpinel this is a bug . It should not be redirecting to https locally.

**Comment by fabiosantoscode** [comment]
@jorgeorpinel NODE_ENV is set to production, making `dev` false, which makes the redirect stuff issue a redirect to HTTPS. I'll send a PR to avoid such a redirect on localhost.

**Comment by shcheklein** [comment]
@fabiosantoscode I would check if the redirects engine itself also matches on https only (thus preventing us from testing at least some redirects on localhost)

**Comment by fabiosantoscode** [comment]
@shcheklein path redirects are relative, so they go to localhost. Hostname redirects (man.dvc.org, error.dvc.org) take you to the real site, where HTTPS works.

File: server.js [hunk 1]
```diff
@@ -23,43 +24,38 @@ const port = process.env.PORT || 3000
 
 app.prepare().then(() => {
   createServer((req, res) => {
-    const parsedUrl = parse(req.url, true)
-    const { pathname, query } = parsedUrl
+    const parsedUrl = parseURL(req.url)
+    const { pathname, queryStr } = parsedUrl
```
Comment (fabiosantoscode) [inline server.js]: There's no `queryStr` when parsing a URL. The querystring is called "query" just the same. The difference is that it's an object if `url.parse`'s second argument is true.

File: src/utils/redirects.test.js [hunk 2]
```diff
@@ -3,41 +3,41 @@ const { processRedirectString, getRedirect } = require('./redirects')
...
 describe('processRedirectString', () => {
   it('reads the regex, replacement and code', () => {
-    const { regex, replace, code } = processRedirectString('^/foo /bar 418')
+    const { regex, replace, code } = processRedirectString('^/foo /bar 308')
 
     expect(regex).toBeInstanceOf(RegExp)
     expect(regex.source).toEqual('^\\/foo')
     expect(replace).toEqual('/bar')
-    expect(code).toEqual(418)
+    expect(code).toEqual(308)
```
Comment (fabiosantoscode) [inline src/utils/redirects.test.js]: You removed my teapot joke :( it's old anyway.

File: server.js (hunk 1 above)
Comment (jorgeorpinel) [reply server.js]: Thanks! Yes, this is part of the problem (my bad, fixed in b85eb64). But the actual issue is a bug in Next, it seems. See https://github.com/zeit/next.js/issues/7303#issuecomment-511053550

Addressing... ⏳

File: src/utils/redirects.test.js (hunk 2 above)
Comment (jorgeorpinel) [reply src/utils/redirects.test.js]: Wow I didn't even remember that one. I thought it was just a random number haha. Reverted now.

**Comment by jorgeorpinel** [comment]
Thanks Fabio, that was part of it.

Fixed this now in e51bc64 ! It's actually a kinda bug from Next... It expects that you always call `url.parse` sending `true` to its 2nd param (`parseQueryString`)... See https://github.com/zeit/next.js/issues/7303#issuecomment-589267045

**Comment by jorgeorpinel** [comment]
In any case... This review app works well now. https://dvc-landing-server-redi-btalyj.herokuapp.com/

Thanks everyone for all the feedback.

File: server.js [hunk 3]
```diff
@@ -23,43 +24,39 @@ const port = process.env.PORT || 3000
...
         res.statusCode = 404
       }
 
       // Custom route for all docs
-      app.render(req, res, '/doc', query)
+      app.render(req, res, '/doc', parseQuery(queryStr))
     } else {
       // Regular Next.js handler
+
+      if (null === parsedUrl.query) parsedUrl.query = {}
```
Comment (shcheklein) [inline server.js]: does it worth doing this minor optimization considering that we have to this hack?

Comment (jorgeorpinel) [reply server.js]: Probably not.

I don't see it as a hack though, just part of the necessary custom Next server logic.

But anyway should I just revert this whole file and leave the redirect file changes (totally cosmetic)? If this whole PR is pointless at this point, please just close it 🙂

Comment (shcheklein) [reply server.js]: up to you @jorgeorpinel  :)

Comment (jorgeorpinel) [reply server.js]: OK I reverted the optimization and hack. But then I found a better one! Implemented in 08522d88, please check.

**Comment by shcheklein** [comment]
👍 LGTM

**Comment by jorgeorpinel** [comment]
Deployed and double sanity checked on prod. All good! 🎉

**Comment by fabiosantoscode** [comment]
@jorgeorpinel let me know if you had issues testing redirects locally. I tested them with:

`curl -I --header 'host: man.dvc.org` localhost:3000/something`

And checked the status code and location header for each redirect. This is also possible with HTTP GUIs like postman.

**Comment by jorgeorpinel** [comment]
Seems to work properly after #1010. Thanks!

**Pull request merged**